flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,247 @@
+ * }}}
+ *
+ * =Parameters=
+ *
+ * - [[org.apache.flink.ml.clustering.KMeans.NumIterations]]:
+ * Defines the number of iterations to recalculate the centroids of the clusters. As it
+ * is a heuristic algorithm, there is no guarantee that it will converge to the global optimum. The
+ * centroids of the clusters and the reassignment of the data points will be repeated till the
+ * given number of iterations is reached.
+ * (Default value: '''10''')
+ *
+ * - [[org.apache.flink.ml.clustering.KMeans.InitialCentroids]]:
+ * Defines the initial k centroids of the k clusters. They are used as start off point of the
+ * algorithm for clustering the data set. The centroids are recalculated as often as set in
+ * [[org.apache.flink.ml.clustering.KMeans.NumIterations]]. The choice of the initial centroids
+ * mainly affects the outcome of the algorithm.
+ *
+ */
+class KMeans extends Predictor[KMeans] {
+
+  import KMeans._
+
+  /** Stores the learned clusters after the fit operation */
+  var centroids: Option[DataSet[LabeledVector]] = None
+
+  /**
+   * Sets the number of iterations.
+   *
+   * @param numIterations
+   * @return itself
+   */
+  def setNumIterations(numIterations: Int): KMeans = {
+    parameters.add(NumIterations, numIterations)
+    this
+  }
+
+  /**
+   * Sets the initial centroids on which the algorithm will start computing.
+   * These points should depend on the data and significantly influence the resulting centroids.
+   *
+   * @param initialCentroids A sequence of labeled vectors.
+   * @return itself
+   */
+  def setInitialCentroids(initialCentroids: DataSet[LabeledVector]): KMeans = {
+    parameters.add(InitialCentroids, initialCentroids)
+    this
+  }
+
+}
+
+/**
+ * Companion object of KMeans. Contains convenience functions, the parameter type definitions
+ * of the algorithm and the [[FitOperation]] & [[PredictOperation]].
+ */
+object KMeans {
+  val CENTROIDS = "centroids"
+
+  case object NumIterations extends Parameter[Int] {
+    val defaultValue = Some(10)
+  }
+
+  case object InitialCentroids extends Parameter[DataSet[LabeledVector]] {
+    val defaultValue = None
+  }
+
+  // ========================================== Factory methods ====================================
+
+  def apply(): KMeans = {
+    new KMeans()
+  }
+
+  // ========================================== Operations =========================================
+
+  /**
+   * [[PredictOperation]] for vector types. The result type is a [[LabeledVector]].
+   */
+  implicit def predictValues = {
+    new PredictOperation[KMeans, Vector, LabeledVector] {
+      override def predict(
+        instance: KMeans,
+        predictParameters: ParameterMap,
+        input: DataSet[Vector])
+      : DataSet[LabeledVector] = {
+
+        instance.centroids match {
+          case Some(centroids) => {
+            input.map(new SelectNearestCenterMapper).withBroadcastSet(centroids, CENTROIDS)
```

Comment:
This mapping operation can be replaced by using the new mapWtihBcVariable function. You can check out the function SGDStep in flink.ml.optimization.GradientDescent on how to use it.

It should make code more concise and readable.